src/tag_gc.rs

```diff
@@ -71,6 +71,22 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriInterpCxExt<'mir, 'tcx> {
                         tags.insert(*sb);
                     }
                 }
+                let stacks = alloc
+                    .extra
+                    .stacked_borrows
+                    .as_ref()
+                    .expect("we should not even enter the GC if Stacked Borrows is disabled");
+                tags.extend(&stacks.borrow().exposed_tags);
+
+                if let Some(store_buffers) = alloc.extra.weak_memory.as_ref() {
+                    store_buffers.iter(|val| {
+                        if let Scalar::Ptr(ptr, _) = val {
+                            if let Provenance::Concrete { sb, .. } = ptr.provenance {
+                                tags.insert(sb);
+                            }
+                        }
+                    });
+                }
```

Comment:
I think this is exactly the kind of code you were trying to get rid of with the visitor, but I'm really not sure what to do here because we're implementing the visitor.

Reply:
Hm, yeah allocations are a bit tricky to handle together with the rest of this.

We could have a `VisitProvenance` trait and implement that for `Allocation` and `AllocExtra` and recursively their fields, like we do for machine values?